subsys/settings/src/settings_zms.c

```diff
@@ -0,0 +1,434 @@
+
+	rd_fn_arg = (struct settings_zms_read_fn_arg *)back_end;
+
+	return zms_read(rd_fn_arg->fs, rd_fn_arg->id, data, len);
+}
+
+int settings_zms_src(struct settings_zms *cf)
+{
+	cf->cf_store.cs_itf = &settings_zms_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_zms_dst(struct settings_zms *cf)
+{
+	cf->cf_store.cs_itf = &settings_zms_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_zms_update_ll(struct settings_zms *cf, uint32_t name_hash)
+{
+	int rc = 0;
+	struct settings_hash_linked_list settings_element;
+	struct settings_hash_linked_list settings_update_element;
+
+	/* let's update the linked list */
+	rc = zms_read(&cf->cf_zms, name_hash | 1, &settings_element,
+		      sizeof(struct settings_hash_linked_list));
+	if (rc < 0) {
+		return rc;
+	}
+	/* update the next element */
+	if (settings_element.next_hash) {
+		rc = zms_read(&cf->cf_zms, settings_element.next_hash, &settings_update_element,
+			      sizeof(struct settings_hash_linked_list));
+		if (rc < 0) {
+			return rc;
+		}
+		settings_update_element.previous_hash = settings_element.previous_hash;
+		rc = zms_write(&cf->cf_zms, settings_element.next_hash, &settings_update_element,
+			       sizeof(struct settings_hash_linked_list));
+		if (rc < 0) {
+			return rc;
+		}
+		if (!settings_update_element.next_hash) {
+			/* update second_to_last_hash_id */
+			cf->second_to_last_hash_id = settings_element.previous_hash;
+		}
+	} else {
+		/* we are deleting the last element of the linked list
+		 * let's update the last_hash_id.
+		 */
+		cf->last_hash_id = settings_element.previous_hash;
+	}
+	/* update the previous element */
+	if (settings_element.previous_hash) {
+		rc = zms_read(&cf->cf_zms, settings_element.previous_hash, &settings_update_element,
+			      sizeof(struct settings_hash_linked_list));
+		if (rc < 0) {
+			return rc;
+		}
+		if (!settings_element.next_hash) {
+			/* we are deleting the last element of the linked list,
+			 * let's update the second_to_last_hash_id
+			 */
+			cf->second_to_last_hash_id = settings_update_element.previous_hash;
+		}
+		settings_update_element.next_hash = settings_element.next_hash;
+		rc = zms_write(&cf->cf_zms, settings_element.previous_hash,
+			       &settings_update_element, sizeof(struct settings_hash_linked_list));
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+static int settings_zms_delete(struct settings_zms *cf, uint32_t name_hash)
+{
+	int rc = 0;
+
+	rc = zms_delete(&cf->cf_zms, name_hash);
+	if (rc >= 0) {
+		rc = zms_delete(&cf->cf_zms, name_hash + ZMS_DATA_ID_OFFSET);
+	}
+	if (rc < 0) {
+		return rc;
+	}
+
+	rc = settings_zms_update_ll(cf, name_hash);
+	if (rc < 0) {
+		return rc;
+	}
+
+	/* Now delete the current linked list element */
+	rc = zms_delete(&cf->cf_zms, name_hash | 1);
+
+	return rc;
+}
+
+static int settings_zms_load(struct settings_store *cs, const struct settings_load_arg *arg)
```

Comment:
You seem to need extra reads compared to settings over NVS because of traversing the hash list. Just note that protocol stacks that utilize settings often do not read all settings at once but call `settings_load_subtree_direct`. For this reason, the performance of `settings_load_subtree_direct` is actually more or equally important for them than the performance of `settings_save`. I suggest adding a bechmark for loading individual keys as well to validate there's no performance impact for this use case.

Maybe this use case could be optimized as well by first loading the key that matches `arg->subtree` (using the hash and the same key lookup as `settings_save_one` does) and only then the remaining ones (which gives the application the opportunity to stop after the first key if it actually meant to load this key and not the whole subtree).

Reply:
The settings over NVS needs also some additional reads. Actually it needs to find the name_id by going through the whole name_id list (if entry is not in cache) before reading the right path and data.
I will have a look at the settings_load_subtree_direct to see what it calls

Reply:
Yeah, I meant additional compared to settings over NVS, which doesn't require a read to move on to the next key.